### ISSUE TITLE
Bug in spec resolution of virtual dependencies

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#3764]
I'm trying to build `hpl`, which contains:
```python
depends_on('mpi@1.1:')                                                      
depends_on('blas')
```
The concretized spec looks like:
```
$ spack spec hpl
...
hpl@2.2%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX" +openmp arch=linux-centos6-x86_64 
    ^intel-parallel-studio@professional.2017.1%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX" +all+daal~ilp64+ipp+mkl~mpi~newdtags+openmp+rpath+shared+tools arch=linux-centos6-x86_64 
    ^mvapich2@2.2%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX" ~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock arch=linux-centos6-x86_64 
        ^bison@2.4.1%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX"  arch=linux-centos6-x86_64 
        ^libpciaccess@0.13.4%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX"  arch=linux-centos6-x86_64 
            ^libtool@2.2.6b%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX"  arch=linux-centos6-x86_64 
            ^pkg-config@0.23%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX" +internal_glib arch=linux-centos6-x86_64 
            ^util-macros@1.19.1%intel@17.0.2 cflags="-axCORE-AVX512,CORE-AVX2,AVX" cxxflags="-axCORE-AVX512,CORE-AVX2,AVX" fflags="-axCORE-AVX512,CORE-AVX2,AVX"  arch=linux-centos6-x86_64
```
However, `spack install` crashes with the following error message:
```
==> Error: AttributeError: 'Spec' object has no attribute 'mpicc'
```
At first, I thought this was another case of #3487, but that's where things get weird. `mvapich2` defines `self.mpicc`, so I was confused why I would get this error message. On a whim, I added:
```python
print(spec['mpi'].name)
```
to the `hpl` package, only to find that it returns `intel-parallel-studio`.

Now, my `intel-parallel-studio` installation is the `professional` version, so it doesn't come with Intel MPI. I didn't want to reinstall such a large package every time I built something that required BLAS/LAPACK, so I installed it with Spack and added it to my `packages.yaml` file:
```yaml
packages:
all:
    compiler: [intel, gcc, pgi, nag]
    providers:
      blas:   [intel-parallel-studio]
      lapack: [intel-parallel-studio]
      mpi: [mvapich2, mpich, openmpi]
  intel-parallel-studio:
    paths:
      intel-parallel-studio@professional.2017.1+all+daal~ilp64+ipp+mkl~mpi~newdtags+openmp+rpath+shared+tools arch=linux-centos6-x86_64: /soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4
    buildable: False
```
Here I'm clearly setting `~mpi` because it doesn't provide MPI. Yet for some reason, `spec['mpi']` resolves to `spec['intel-parallel-studio']`. And for some reason, `spack spec hpl` correctly adds `mvapich2` to the list of dependencies.

My theory is that there's a bug in the way we specify provides in the `intel-parallel-studio` package:
```python
provides('mpi', when='@cluster:+mpi')
```
I have a feeling that `@professional.2017.1` satisfies `@cluster:`. But I don't see why it would satisfy `+mpi` since I clearly marked it as `~mpi` in my `packages.yaml`. And I don't understand why `spack spec` would correctly choose `mvapich2` but `spec['mpi']` wouldn't.

By the way, I tried:
```
$ spack install hpl ^openblas
```
and that succeeded, so it's definitely a problem with `intel-parallel-studio` being mistakenly picked up as the MPI provider.

@tgamblin @alalazo you may want to look at this.

**Comment by adamjstewart** [comment]
For now, commenting out the `provides('mpi')` directive in `intel-parallel-studio` lets me build `hpl`, although there are some other build errors I need to figure out.

**Comment by adamjstewart** [comment]
The build error I mentioned looks like:
```
/blues/gpfs/home/ajstewart/spack/opt/spack/linux-centos6-x86_64/intel-17.0.2/mvapich2-2.2-tnt7l6y7eazcwrzzgdwxc3gzuxfnvqns/bin/mpicc -DAdd_ -DF77_INTEGER=int -DStringSunStyle -DHPL_DETAILED_TIMING -DHPL_PROGRESS_REPORT -I/scratch/ajstewart/spack-stage/spack-stage-pNaqVO/hpl-2.2/include -I/scratch/ajstewart/spack-stage/spack-stage-pNaqVO/hpl-2.2/include/Linux-x86_64 -I/soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4/include -I-I/blues/gpfs/home/ajstewart/spack/opt/spack/linux-centos6-x86_64/intel-17.0.2/mvapich2-2.2-tnt7l6y7eazcwrzzgdwxc3gzuxfnvqns/include -O3 -qopenmp -o /scratch/ajstewart/spack-stage/spack-stage-pNaqVO/hpl-2.2/bin/Linux-x86_64/xhpl HPL_pddriver.o         HPL_pdinfo.o           HPL_pdtest.o /scratch/ajstewart/spack-stage/spack-stage-pNaqVO/hpl-2.2/lib/Linux-x86_64/libhpl.a /soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4/mkl/lib/intel64/libmkl_intel_lp64.so /soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4/mkl/lib/intel64/libmkl_core.so /soft/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-4.4.7/intel-parallel-studio-professional.2017.1-zaq5tbp664b2ye2gdx3ry35t3wd4rjn4/mkl/lib/intel64/libmkl_intel_thread.so libpthread.so libm.so libdl.so -L/blues/gpfs/home/ajstewart/spack/opt/spack/linux-centos6-x86_64/intel-17.0.2/mvapich2-2.2-tnt7l6y7eazcwrzzgdwxc3gzuxfnvqns/lib
icc: error #10236: File not found:  'libpthread.so'
icc: error #10236: File not found:  'libm.so'
icc: error #10236: File not found:  'libdl.so'
```
In the `intel-parallel-studio` package, it looks like we're adding these last 3 system libraries by hand:
```python
system_libs = [                                                         
    'libpthread.{0}'.format(suffix),                                    
    'libm.{0}'.format(suffix),                                          
    'libdl.{0}'.format(suffix)                                          
]                                                                       
return mkl_libs + system_libs
```
Does anyone know why we were doing this? @lee218llnl @davydden @alalazo

**Comment by davydden** [comment]
> Does anyone know why we were doing this? 

because if you look at Intel MKL link adviser, you will see a need to provide all those libs. Otherwise MKL won't work in some packages

**Comment by adamjstewart** [comment]
Won't those libraries be automatically picked up from the system anyway? If not, then we need to add the full path, not just the file name.

**Comment by davydden** [comment]
> Won't those libraries be automatically picked up from the system anyway?

no, they won't. You will get undefined symbols during linking.

**Comment by davydden** [comment]
p.s. @adamjstewart here is the link line from  https://software.intel.com/en-us/articles/intel-mkl-link-line-advisor  for intel compiler
```
 -L${MKLROOT}/lib/intel64 -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```
As you see, we need to add those system libs.

But I agree, probably we need to specify their full path to make sure `LibraryList` works correctly for all usage scenarios.

**Comment by alalazo** [comment]
Reading this bug, and it seems like it might be due to an issue that has been fixed in #7716 Can you check?

**Comment by adamjstewart** [comment]
I no longer have `intel-parallel-studio~mpi` installed, so this is difficult to test. But I suspect that you're right. I'm going to close this. If someone else runs into the same problem, feel free to reopen.
